stm32-modules/include/thermocycler-refresh/thermocycler-refresh/tmc2130.hpp

```diff
@@ -0,0 +1,316 @@
+     */
+    [[nodiscard]] auto get_cool_threshold() const -> TCoolThreshold {
+        return _registers.tcoolthrs;
+    }
+
+    /**
+     * @brief Update THIGH register
+     * @param reg New configuration register to set
+     * @param policy Instance of abstraction policy to use
+     * @return True if new register was set succesfully, false otherwise
+     */
+    template <TMC2130Policy Policy>
+    auto set_thigh(THigh reg, Policy& policy) -> bool {
+        if (set_register(policy, reg)) {
+            _registers.thigh = reg;
+            return true;
+        }
+        return false;
+    }
+    /**
+     * @brief Get the current THIGH register status
+     */
+    [[nodiscard]] auto get_thigh() const -> THigh { return _registers.thigh; }
+
+    /**
+     * @brief Update CHOPCONF register
+     * @param reg New configuration register to set
+     * @param policy Instance of abstraction policy to use
+     * @return True if new register was set succesfully, false otherwise
+     */
+    template <TMC2130Policy Policy>
+    auto set_chop_config(ChopConfig reg, Policy& policy) -> bool {
+        if (set_register(policy, reg)) {
+            _registers.chopconf = reg;
+            return true;
+        }
+        return false;
+    }
+    /**
+     * @brief Get the current CHOPCONF register status
+     */
+    [[nodiscard]] auto get_chop_config() const -> ChopConfig {
+        return _registers.chopconf;
+    }
+
+    /**
+     * @brief Update COOLCONF register
+     * @param reg New configuration register to set
+     * @param policy Instance of abstraction policy to use
+     * @return True if new register was set succesfully, false otherwise
+     */
+    template <TMC2130Policy Policy>
+    auto set_cool_config(CoolConfig reg, Policy& policy) -> bool {
+        // Assert that bits that MUST be 0 are actually 0
+        if ((reg.padding_1 != 0) || (reg.padding_2 != 0) ||
+            (reg.padding_3 != 0) || (reg.padding_4 != 0)) {
+            return false;
+        }
+        if (set_register(policy, reg)) {
+            _registers.coolconf = reg;
+            return true;
+        }
+        return false;
+    }
+    /**
+     * @brief Get the current COOLCONF register status
+     */
+    [[nodiscard]] auto get_cool_config() const -> CoolConfig {
+        return _registers.coolconf;
+    }
+
+  private:
+    /**
+     * @brief Set a register on the TMC2130
+     *
+     * @tparam Reg The type of register to set
+     * @tparam Policy Abstraction class for actual writing
+     * @param[in] policy Instance of the abstraction policy to use
+     * @param[in] reg The register to write
+     * @return True if the register could be written, false otherwise.
+     * Attempts to write to an unwirteable register will throw a static
+     * assertion.
+     */
+    template <TMC2130Register Reg, TMC2130Policy Policy>
+    auto set_register(Policy& policy, Reg reg) -> bool {
+        static_assert(Reg::writable);
```

Comment:
this could also be 
```
template <TMC2130Register Reg, TMC2130Policy Policy>
requires (const Reg& reg) { {reg::writable }; }
auto set_register(...
```
i believe. Doing it this way might give better error messages, and would be consistent with our use of requirements elsewhere.

I think you could also use a named-requirement or concept for this, possibly by making a `WritableTMC2130Register` and `ReadableTMC2130Register` that looks like the requirement above `&& TMC2130Register`

Reply:
Hmm, tried this and it doesn't work as expected - you can read from a nonreadable register and write to a nonwritable register. I think this just checks that the variable exists - looking through the `concepts` library I'm not sure what the right syntax would be to check that the value of the variable is `true`.

Reply:
There's a way to do this I think maybe but the better way is something like this, from the "type requirements" section of https://en.cppreference.com/w/cpp/language/constraints (no section anchors, bad cppreference). These can also be `using`-declarations. You can also do this with straight up old style SFINAE using `enable_if` but that's getting to be a little much I think

```c++
template <typename _RegType>
concept WritableRegister = requires {
    typename _RegType::writable; 
}

template <typename _RegType>
concept ReadableRegister = requires {
    typename _RegType::readable; 
}

struct SomeRWRegister {
   static constexpr bool writable = true;
   static constexpr bool readable = true;
};

struct SomeRORegister {
   static constexpr bool readable = true;
   // do not define writable
};

struct SomeWORegister {
  static constexpr bool writable = true;
  // do not define readable
};
```

Reply:
Ahh makes sense, I wasn't getting rid of the `false` flags